src/compare.c

```diff
@@ -1,136 +1,325 @@
+      stop_not_comparable(x, y, "data frame with zero columns");
+    }
+
+    return df_compare_scalar(x, i, y, j, na_equal, n_col);
   }
+  default: break;
+  }
+
+  Rf_errorcall(R_NilValue, "Unsupported type %s", Rf_type2char(TYPEOF(x)));
 }
 
+// -----------------------------------------------------------------------------
+
+static SEXP df_compare(SEXP x, SEXP y, bool na_equal, R_len_t n);
+
+#define COMPARE(CTYPE, CONST_DEREF, SCALAR_COMPARE)     \
+do {                                                    \
+  SEXP out = PROTECT(Rf_allocVector(INTSXP, n));        \
+  int* p_out = INTEGER(out);                            \
+                                                        \
+  const CTYPE* p_x = CONST_DEREF(x);                    \
+  const CTYPE* p_y = CONST_DEREF(y);                    \
+                                                        \
+  for (R_len_t i = 0; i < n; ++i, ++p_x, ++p_y) {       \
+    p_out[i] = SCALAR_COMPARE(p_x, p_y, na_equal);      \
+  }                                                     \
+                                                        \
+  UNPROTECT(1);                                         \
+  return out;                                           \
+}                                                       \
+while (0)
+
 // [[ register() ]]
-SEXP vctrs_compare(SEXP x, SEXP y, SEXP na_compare_) {
-  bool na_compare = Rf_asLogical(na_compare_);
+SEXP vctrs_compare(SEXP x, SEXP y, SEXP na_equal_) {
+  bool na_equal = Rf_asLogical(na_equal_);
+
   R_len_t n = vec_size(x);
 
-  SEXP out = PROTECT(Rf_allocVector(INTSXP, n));
-  int32_t* p_out = INTEGER(out);
+  enum vctrs_type type = vec_proxy_typeof(x);
+  if (type != vec_proxy_typeof(y) || n != vec_size(y)) {
+    stop_not_comparable(x, y, "must have the same types and lengths");
+  }
 
-  for (R_len_t i = 0; i < n; ++i) {
-    p_out[i] = compare_scalar(x, i, y, i, na_compare);
+  switch (type) {
+  case vctrs_type_logical:   COMPARE(int, LOGICAL_RO, lgl_compare_scalar);
+  case vctrs_type_integer:   COMPARE(int, INTEGER_RO, int_compare_scalar);
+  case vctrs_type_double:    COMPARE(double, REAL_RO, dbl_compare_scalar);
+  case vctrs_type_character: COMPARE(SEXP, STRING_PTR_RO, chr_compare_scalar);
+  case vctrs_type_dataframe: return df_compare(x, y, na_equal, n);
+  case vctrs_type_scalar:    Rf_errorcall(R_NilValue, "Can't compare scalars with `vctrs_compare()`");
+  case vctrs_type_list:      Rf_errorcall(R_NilValue, "Can't compare lists with `vctrs_compare()`");
+  default:                   Rf_error("Unimplemented type in `vctrs_compare()`");
   }
+}
+
+#undef COMPARE
+
+// -----------------------------------------------------------------------------
+
+static SEXP vec_compare_col(SEXP x, SEXP y, bool na_equal, SEXP info, R_len_t n);
+static SEXP df_compare_impl(SEXP x, SEXP y, bool na_equal, SEXP info, R_len_t n);
+
+static SEXP df_compare(SEXP x, SEXP y, bool na_equal, R_len_t n) {
+  // Skip nothing to begin with
```

Comment:
It would help to explain the algorithm here, e.g. what `skip` is for and what `count` is counting.

Maybe `count` should be `remaining_rows` and `skip` something like `row_done` to indicate it's about rows that are already compared?